ports/stm32f3/board_flash.c

```diff
@@ -71,59 +71,76 @@ static bool is_blank(uint32_t addr, uint32_t size)
   return true;
 }
 
-                    //ADDR            data
-void flash_write(uint32_t dst, const uint8_t *src, int len)
+static bool flash_erase(uint32_t addr)
 {
-   // assume sector 0-7 (bootloader) is same size as sector 1
-  uint32_t addr = APP_LOAD_ADDRESS;
+  // starting address from 0x08000000
+  uint32_t sector_addr = FLASH_BASE_ADDR;
+  bool erased = false;
+
   uint32_t sector = 0;
-  int erased = false;
   uint32_t size = 0;
 
-  for ( unsigned i = 0; i < BOARD_FLASH_SECTORS; i++ )
+  for ( uint32_t i = 0; i < SECTOR_COUNT; i++ )
   {
+    TUF2_ASSERT(sector_addr < FLASH_BASE_ADDR + BOARD_FLASH_SIZE);
+
     size = flash_func_sector_size(i);
-    if ( addr + size > dst )
+    if ( sector_addr + size > addr )
     {
-      sector = i + 1;
-      erased = erasedSectors[i];
-      erasedSectors[i] = 1;    // don't erase anymore - we will continue writing here!
+      sector = i;
+      erased = erased_sectors[i];
+      erased_sectors[i] = 1;    // don't erase anymore - we will continue writing here!
       break;
     }
-    addr += size;
+    sector_addr += size;
   }
 
-  if (sector == 0)
-  {
-    TUF2_LOG1("invalid sector\r\n");
-  }
+  (void)sector;
 
-  HAL_FLASH_Unlock();
+#ifndef TINYUF2_SELF_UPDATE
+  // skip erasing sector0 if not self-update
+  TUF2_ASSERT(sector);
+#endif
 
-  if (!erased && !is_blank(addr, size))
+  if ( !erased && !is_blank(sector_addr, size))
   {
-    uint32_t SectorError = 0;
-
-    TUF2_LOG1("Erase: %08lX size = %lu\n", addr, size);
+    TUF2_LOG1("Erase: %08lX size = %lu KB ... ", sector_addr, size / 1024);
 
     FLASH_EraseInitTypeDef EraseInit;
     EraseInit.TypeErase = FLASH_TYPEERASE_PAGES;
-    EraseInit.PageAddress = addr;
-    EraseInit.NbPages = ((0x08040000 - addr)/size);
+    EraseInit.PageAddress = sector_addr;
+    EraseInit.NbPages = ((BOARD_FLASH_APP_START - sector_addr)/size);
```

Comment:
I know this is an historical value, but looks like `EraseInit.NbPages` should always be 1. sector_addr can be larger than the `BOARD_FLASH_APP_START` when programming an actual app, I have no idea why I put this value in the first place. Would you mind doing a quick check to comfirm this.

Reply:
Number of pages would be 8 on the default start address (16K of 2K pages). However I have code locally that adds additional board configuration, which currently goes over that 16K marker and would want to keep this area somewhat dynamic so that its less areas to update.

Reply:
the flash_erase() is called per each sector, so I think this should just be 1, yes ?

Reply:
I want to say `flash_write` is called with chunks larger than a page, though I will have to give this a proper look when im back at a machine.

Reply:
look good, let me make the changes.

Reply:
As long as there is the assumption that `flash_write` does not get called with an address+length that would cross 2 flash pages, I think the code should be okay.

Reply:
for now, it won't do that, so I guess we are good (for now)